### ISSUE TITLE
pls support tree style tab

### ISSUE THREAD

**Issue by perdolka** [rNeomy/auto-tab-discard#11]
TST: https://github.com/piroor/treestyletab
**API for other addons:** https://github.com/piroor/treestyletab/wiki/API-for-other-addons

**Comment by philipbel** [comment]
+1

**Comment by rNeomy** [comment]
What is the issue with the "tree style tab" extension?

**Comment by perdolka** [comment]
The auto-tab-discard options do not appear in the right-click menu on tabs.

**Comment by rNeomy** [comment]
So we have the options on the right-click menu over the tab context and the toolbar button context. Do we really need them on the sidebar too?

**Comment by perdolka** [comment]
Most of TST-users hide default horizontal tabs (via UserChrome.css).
So, usually there is only toolbar button context for them.
But at the same time, many are accustomed to manage tabs directly through the RMB-context on them.

Maybe its really isn't necessary, just my opinion.

**Comment by rNeomy** [comment]
okay then, let's keep this open for a pull merge to support TST integration. Ideally I prefer to call something like `app.contextMenus.create` to create all the contexts and `chrome.contextMenus.onClicked.addListener` for callbacks.

**Comment by ssokolow** [comment]
As explained in the [Add new item to the context menu on tabs](https://github.com/piroor/treestyletab/wiki/API-for-other-addons#add-new-item-to-the-context-menu-on-tabs) section of the TST documentation that was linked...

> You can share and reuse parameters for your context menu to both Firefox's menu and TST's menu. 
>
> For example:
> 
> ```js
> let params = {
>   id, type, title, parentId,
>   contexts: ['page', 'tab']
> };
> await browser.menus.create(params);
> await browser.runtime.sendMessage(kTST_ID, {
>   type: 'fake-contextMenu-create',
>   params
> }).catch(error => { /* TST is not available */ });
> ```

(`kTST_ID` is defined in a previous example as `const kTST_ID = 'treestyletab@piro.sakura.ne.jp';`)

Beyond that, we're waiting on Mozilla:

> Due to limitations of WebExtensions APIs, context menu on tabs is not available on custom tab bar in the sidebar panel. (See also [bug 1280347](https://bugzilla.mozilla.org/show_bug.cgi?id=1280347), [bug 1376251](https://bugzilla.mozilla.org/show_bug.cgi?id=1376251), and [bug 1396031](https://bugzilla.mozilla.org/show_bug.cgi?id=1396031).) Thus TST 2.0 provides its custom context menu in the sidebar. Custom menu items added by browser.menus.create() don't appear in the menu because it is just a fake. Until any of bugs I listed above become fixed, your addon need to register its custom menu item to TST's fake context menu manually.
>
> -- [source](https://github.com/piroor/treestyletab/wiki/API-for-other-addons#extra-context-menu-items-on-tabs)

**Comment by rNeomy** [comment]
Thanks @ssokolow. We have TST integration on 0.1.5

**Comment by perdolka** [comment]
@rNeomy @ssokolow  Thank you!

**Comment by Walkman100** [comment]
Thanks very much for the context menu, it works really well. would it be possible (or in scope of this addon) to add a "Discard this tab tree" option on right-clicking a tree of tabs? I don't know if TST exposes functionality that would allow this, or if there is some reason why it might be a bad idea (unloading too many tabs at once or something)

**Comment by lbmaian** [comment]
Thanks for the TST support.

However, there is an issue where reinstalling TST (which also happens when upgrading TST) "deregisters" Auto Tab Discard in the TST context menu. I have to disable and re-enable Auto Tab Discard in about:addons to get Auto Tab Discard to reappear in the TST context menu.

I don't see this problem with other TST addons like https://github.com/joshuacant/tst_open_in_private

**Comment by rNeomy** [comment]
@Walkman100 the latest dev version has a new menu item to discard the entire tree.
@lbmaian Fixes on the latest dev version

Please check and report issues.

**Comment by lbmaian** [comment]
I just tested the new version out. The TST fake submenu registering worked once, when I installed the latest nightly version of TST, but reinstalling that version or rolling back to the AMO version (and reinstalling that) resulted in the TST fake submenu disappearing.

To be fair, this also affects the TST addon that I linked above, so this is probably a problem on TST's side. In any case, the TST fake menu is hack that hopefully won't be needed in future versions of Firefox, and the workaround of toggling off and on Auto Tab Discard works fine.

Thanks for all your work!
